File: androidApp/src/main/kotlin/com/appelis/kmp_demo/android/screens/CategoryView.kt
package com.appelis.kmp_demo.android.screens

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import com.appelis.kmp_demo.assortment.uiLogic.category.CategoryComponent
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.LineHeightStyle
import androidx.compose.ui.unit.dp
import com.appelis.kmp_demo.assortment.uiLogic.category.CategoryViewState

@Composable
fun CategoryView(
    component: CategoryComponent
) {
    val viewModel = remember { component.viewModel }
    val viewState by viewModel.viewState.collectAsState()

    LaunchedEffect(Unit) {
        viewModel.setup()
    }

    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier.fillMaxSize()
    ) {
        when (viewState) {
            is CategoryViewState.ArticleCollection -> CategoryArticleCollectionView(
                component = component.articleCollectionSubComponent,
                categoryId = (viewState as? CategoryViewState.ArticleCollection)?.parentCategory?.id ?: ""
            )
            is CategoryViewState.CategoryCollection -> CategoryCollectionView(
                component = component.categoryCollectionSubComponent,
                parentId = (viewState as? CategoryViewState.CategoryCollection)?.parentCategory?.id  ?: ""
            )
            CategoryViewState.GeneralError -> GeneralError()
            CategoryViewState.Loading -> Loading()
            CategoryViewState.NetworkError -> NetworkError()
        }
    }
}

@Composable
private fun Loading() {
    CircularProgressIndicator()
}

@Composable
private fun NetworkError() {
    Text("Network Error")
}

@Composable
private fun GeneralError() {
    Text("General Error")
}

File: iosApp/Template/Sources/Solaris/Features/Assortment/Category/CategoryView.swift
//
//  CategoryView.swift
//
//
//  Created by Jan Maloušek on 10.06.2024.
//

import Shared
import SwiftUI
import SwiftUICore

struct CategoryView: View {
    @State
    private var viewState: CategoryViewState
    
    private let viewModel: CategoryComponentViewModel
    private let component: CategoryComponent
    
    public init(component: CategoryComponent) {
        self.component = component
        self.viewModel = component.viewModel
        self.viewState = .Loading()
    }
    
    var body: some View {
        VStack {
            switch onEnum(of: viewState) {
            case let .articleCollection(data):
                CategoryArticleCollectionView(component: component.articleCollectionSubComponent, categoryId: data.parentCategory.id)
            case let .categoryCollection(data):
                CategoryCollectionView(component: component.categoryCollectionSubComponent, parentId: data.parentCategory.id)
            case .generalError:
                Text("GeneralError")
            case .loading:
                loading
            case .networkError:
                Text("NetworkError")
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color(\.appBackground))
        .onAppear(first: { viewModel.setup() })
        .task {
            for await state in viewModel.viewState {
                self.viewState = state
            }
        }
    }
    
    private var loading: some View {
        VStack {
            Spacer()
            
            ProgressView()
            
            Spacer()
        }
    }
}
